### PR TITLE
riscv: Use a multiarch triple for clang

### PR DESCRIPTION
**Pull request by sorear**
Much like x86_64, the riscv64 GNU toolchain handles both riscv32 and riscv64 depending on compile flags.

### CODE REVIEW

File: cmake-tool/helpers/check_arch_compiler.cmake [hunk 1]
```diff
@@ -14,7 +14,7 @@ function(check_arch_clang)
     elseif("${KernelSel4Arch}" STREQUAL "aarch64")
         string(REGEX MATCH "^aarch64" correct_triple ${TRIPLE})
     elseif("${KernelSel4Arch}" STREQUAL "riscv32")
-        string(REGEX MATCH "^riscv32" correct_triple ${TRIPLE})
+        string(REGEX MATCH "^riscv64" correct_triple ${TRIPLE})
```
Comment (kent-mcleod) [inline cmake-tool/helpers/check_arch_compiler.cmake:17]: This should encode that both riscv64 and riscv32 toolchains can handle riscv32 architectures but will now reject riscv32 toolchains.  Is it not possible to get a riscv32 toolchain anymore?

Comment (sorear) [reply cmake-tool/helpers/check_arch_compiler.cmake:17]: good catch, it's not the default but it's definitely possible.

**Comment by sorear** [comment]
some more issues I hit that I'd like to document somewhere.  These are _not_ requests for fixes:

If you give the build system a TRIPLE that doesn't correspond exactly to an installed gcc toolchain, you get a very opaque error about CMAKE-AR_NOTFOUND.  It would be nice to improve that but I haven't quite figured out how yet.

Builds require LLVM_TARGETS_TO_BUILD to include X86 even if you're cross compiling seL4, the build system does some sanity checks on the found clang without passing a target option.

The process of picking out a libgcc and crt*.o from a gcc install seems quite brittle (it doesn't work for riscv without a bunch of extra options) and possibly superfluous: llvm builds compiler-rt, there should be some way to use that?

**Comment by branden-data61** [comment]
@sorear: Looks like `cmake-tool/helpers/check_arch_compiler.cmake` needs a restyle, but that is all.

Our styling tools are in this same repo, e.g., `misc/style-cmake.sh`.

**Comment by sorear** [comment]
@branden-data61 are you asking me to do that and update the PR?

**Comment by branden-data61** [comment]
> @branden-data61 are you asking me to do that and update the PR?

Yes.  Sorry for the ambiguity.

**Comment by sorear** [comment]
Got it

**Comment by branden-data61** [comment]
@ssrg-bamboo `test`

**Comment by ssrg-bamboo** [comment]
Hello, I'm a bot! I'll bring this PR into Trustworthy Systems and run some tests

**Comment by ssrg-bamboo** [comment]
## Test results for commit 5a3ff5f5715876300f6430b20b69f5f4c3efd8e6

### Summary

| Test name | Result |
| --- | --- |
| Code Quality - Check licenses - seL4_tools - EXT-sorear-riscv_clang-42 | SUCCESSFUL |
| Code Quality - Reuse License Check - seL4_tools - EXT-sorear-riscv_clang-42 | SUCCESSFUL |
| Code Quality - Signed-off-by - seL4_tools - EXT-sorear-riscv_clang-42 | SUCCESSFUL |
| Code Quality - Style - seL4_tools - EXT-sorear-riscv_clang-42 | FAILED |

### Detailed failure logs

| Project | Stage | Result summary |
| --- | --- | --- |
| Code Quality - Style - seL4_tools - EXT-sorear-riscv_clang-42 | Style check on changed files | 1 of 3 jobs failed |

**Job: style**
Build logs

    07-Aug-2020 14:23:01	Build Code Quality - Style - seL4_tools - EXT-sorear-riscv_clang-42 - style #1 (CQ-SEL4TOOLSSTYLE154-SEL4TOOLSSTYLEPR-1) started building on agent A local agent 2
    07-Aug-2020 14:23:01	I run on the Bamboo server!
    07-Aug-2020 14:23:01	Build working directory is /var/atlassian/application-data/bamboo/xml-data/build-dir/151224373/CQ-SEL4TOOLSSTYLE154-SEL4TOOLSSTYLEPR
    07-Aug-2020 14:23:01	Executing build Code Quality - Style - seL4_tools - EXT-sorear-riscv_clang-42 - style #1 (CQ-SEL4TOOLSSTYLE154-SEL4TOOLSSTYLEPR-1)
    07-Aug-2020 14:23:01	Preparing artifact 'code' for use at /var/atlassian/application-data/bamboo/xml-data/build-dir/151224373/CQ-SEL4TOOLSSTYLE154-SEL4TOOLSSTYLEPR/. (location: .)
    07-Aug-2020 14:23:01	Artifact 'code' prepared in 21.02 ms at /var/atlassian/application-data/bamboo/xml-data/build-dir/151224373/CQ-SEL4TOOLSSTYLE154-SEL4TOOLSSTYLEPR/./code.tar.gz
    07-Aug-2020 14:23:01	Running pre-build action: VCS Version Collector
    07-Aug-2020 14:23:01	Starting task 'Untar code' of type 'com.atlassian.bamboo.plugins.scripttask:task.builder.script'
    07-Aug-2020 14:23:01	+ tar xf code.tar.gz --recursive-unlink
    07-Aug-2020 14:23:01	Finished task 'Untar code' with result: Success
    07-Aug-2020 14:23:01	Starting task 'Set permissions so Bamboo can cleanup' of type 'com.atlassian.bamboo.plugins.scripttask:task.builder.script'
    07-Aug-2020 14:23:01	+ chmod -R o+rwx /var/atlassian/application-data/bamboo/xml-data/build-dir/151224373/CQ-SEL4TOOLSSTYLE154-SEL4TOOLSSTYLEPR
    07-Aug-2020 14:23:01	Finished task 'Set permissions so Bamboo can cleanup' with result: Success
    07-Aug-2020 14:23:01	Starting task 'Check style of changed files' of type 'com.atlassian.bamboo.plugins.scripttask:task.builder.script'
    07-Aug-2020 14:23:01	+ cd seL4_tools
    07-Aug-2020 14:23:01	+ xargs -d '\n' /var/atlassian/application-data/bamboo/xml-data/build-dir/151224373/CQ-SEL4TOOLSSTYLE154-SEL4TOOLSSTYLEPR/seL4_tools/misc/style.sh
    07-Aug-2020 14:23:01	++ git merge-base 5a3ff5f5715876300f6430b20b69f5f4c3efd8e6 ccc38660229116bd26807c7012636c70e1c873f3
    07-Aug-2020 14:23:01	+ git diff --name-only 5a3ff5f5715876300f6430b20b69f5f4c3efd8e6 ccc38660229116bd26807c7012636c70e1c873f3
    07-Aug-2020 14:23:02	+ git diff --exit-code
    07-Aug-2020 14:23:02	diff --git a/cmake-tool/helpers/check_arch_compiler.cmake b/cmake-tool/helpers/check_arch_compiler.cmake
    07-Aug-2020 14:23:02	index 5663b9e..151a5f4 100644
    07-Aug-2020 14:23:02	--- a/cmake-tool/helpers/check_arch_compiler.cmake
    07-Aug-2020 14:23:02	+++ b/cmake-tool/helpers/check_arch_compiler.cmake
    07-Aug-2020 14:23:02	@@ -6,17 +6,53 @@
    07-Aug-2020 14:23:02	 
    07-Aug-2020 14:23:02	 function(check_arch_clang)
    07-Aug-2020 14:23:02	     if("${KernelSel4Arch}" STREQUAL "ia32")
    07-Aug-2020 14:23:02	-        string(REGEX MATCH "^x86_64" correct_triple ${TRIPLE})
    07-Aug-2020 14:23:02	+        string(
    07-Aug-2020 14:23:02	+            REGEX
    07-Aug-2020 14:23:02	+                MATCH
    07-Aug-2020 14:23:02	+                "^x86_64"
    07-Aug-2020 14:23:02	+                correct_triple
    07-Aug-2020 14:23:02	+                ${TRIPLE}
    07-Aug-2020 14:23:02	+        )
    07-Aug-2020 14:23:02	     elseif("${KernelSel4Arch}" STREQUAL "x86_64")
    07-Aug-2020 14:23:02	-        string(REGEX MATCH "^x86_64" correct_triple ${TRIPLE})
    07-Aug-2020 14:23:02	+        string(
    07-Aug-2020 14:23:02	+            REGEX
    07-Aug-2020 14:23:02	+                MATCH
    07-Aug-2020 14:23:02	+                "^x86_64"
    07-Aug-2020 14:23:02	+                correct_triple
    07-Aug-2020 14:23:02	+                ${TRIPLE}
    07-Aug-2020 14:23:02	+        )
    07-Aug-2020 14:23:02	     elseif("${KernelSel4Arch}" STREQUAL "aarch32" OR "${KernelSel4Arch}" STREQUAL "arm_hyp")
    07-Aug-2020 14:23:02	-        string(REGEX MATCH "^arm" correct_triple ${TRIPLE})
    07-Aug-2020 14:23:02	+        string(
    07-Aug-2020 14:23:02	+            REGEX
    07-Aug-2020 14:23:02	+                MATCH
    07-Aug-2020 14:23:02	+                "^arm"
    07-Aug-2020 14:23:02	+                correct_triple
    07-Aug-2020 14:23:02	+                ${TRIPLE}
    07-Aug-2020 14:23:02	+        )
    07-Aug-2020 14:23:02	     elseif("${KernelSel4Arch}" STREQUAL "aarch64")
    07-Aug-2020 14:23:02	-        string(REGEX MATCH "^aarch64" correct_triple ${TRIPLE})
    07-Aug-2020 14:23:02	+        string(
    07-Aug-2020 14:23:02	+            REGEX
    07-Aug-2020 14:23:02	+                MATCH
    07-Aug-2020 14:23:02	+                "^aarch64"
    07-Aug-2020 14:23:02	+                correct_triple
    07-Aug-2020 14:23:02	+                ${TRIPLE}
    07-Aug-2020 14:23:02	+        )
    07-Aug-2020 14:23:02	     elseif("${KernelSel4Arch}" STREQUAL "riscv32")
    07-Aug-2020 14:23:02	-        string(REGEX MATCH "^riscv32" correct_triple ${TRIPLE})
    07-Aug-2020 14:23:02	+        string(
    07-Aug-2020 14:23:02	+            REGEX
    07-Aug-2020 14:23:02	+                MATCH
    07-Aug-2020 14:23:02	+                "^riscv32"
    07-Aug-2020 14:23:02	+                correct_triple
    07-Aug-2020 14:23:02	+                ${TRIPLE}
    07-Aug-2020 14:23:02	+        )
    07-Aug-2020 14:23:02	     elseif("${KernelSel4Arch}" STREQUAL "riscv64")
    07-Aug-2020 14:23:02	-        string(REGEX MATCH "^riscv64" correct_triple ${TRIPLE})
    07-Aug-2020 14:23:02	+        string(
    07-Aug-2020 14:23:02	+            REGEX
    07-Aug-2020 14:23:02	+                MATCH
    07-Aug-2020 14:23:02	+                "^riscv64"
    07-Aug-2020 14:23:02	+                correct_triple
    07-Aug-2020 14:23:02	+                ${TRIPLE}
    07-Aug-2020 14:23:02	+        )
    07-Aug-2020 14:23:02	     else()
    07-Aug-2020 14:23:02	         message(SEND_ERROR "KernelSel4Arch is not set to a valid arch")
    07-Aug-2020 14:23:02	     endif()
    07-Aug-2020 14:23:02	@@ -67,7 +103,7 @@ function(check_arch_gcc)
    07-Aug-2020 14:23:02	 endfunction()
    07-Aug-2020 14:23:02	 
    07-Aug-2020 14:23:02	 function(check_arch_compiler)
    07-Aug-2020 14:23:02	-    if (CMAKE_C_COMPILER_ID STREQUAL "Clang")
    07-Aug-2020 14:23:02	+    if(CMAKE_C_COMPILER_ID STREQUAL "Clang")
    07-Aug-2020 14:23:02	         check_arch_clang()
    07-Aug-2020 14:23:02	     else()
    07-Aug-2020 14:23:02	         check_arch_gcc()
    07-Aug-2020 14:23:02	Failing task since return code of [/var/atlassian/application-data/bamboo/temp/CQ-SEL4TOOLSSTYLE154-SEL4TOOLSSTYLEPR-1-ScriptBuildTask-8446616291750394773.sh] was 1 while expected 0
    07-Aug-2020 14:23:02	Finished task 'Check style of changed files' with result: Failed
    07-Aug-2020 14:23:02	Starting task 'Set permissions so Bamboo can cleanup' of type 'com.atlassian.bamboo.plugins.scripttask:task.builder.script'
    07-Aug-2020 14:23:02	+ chmod -R o+rwx /var/atlassian/application-data/bamboo/xml-data/build-dir/151224373/CQ-SEL4TOOLSSTYLE154-SEL4TOOLSSTYLEPR
    07-Aug-2020 14:23:02	Finished task 'Set permissions so Bamboo can cleanup' with result: Success
    07-Aug-2020 14:23:02	Starting task 'Run astyle code insights' of type 'com.atlassian.bamboo.plugins.scripttask:task.builder.script'
    07-Aug-2020 14:23:02	+ cd seL4_tools
    07-Aug-2020 14:23:02	++ echo ssh://git@bitbucket.ts.data61.csiro.au:7999/sel4/sel4_tools.git
    07-Aug-2020 14:23:02	++ cut -d/ -f4
    07-Aug-2020 14:23:02	+ bb_proj=sel4
    07-Aug-2020 14:23:02	+ /scripts/bb_code_insights/create_code_insights.sh astyle sel4 seL4_tools
    07-Aug-2020 14:23:02	++ dirname /scripts/bb_code_insights/create_code_insights.sh
    07-Aug-2020 14:23:02	+ SCRIPT_DIR=/scripts/bb_code_insights
    07-Aug-2020 14:23:02	+ BBS_URL=https://bitbucket.ts.data61.csiro.au
    07-Aug-2020 14:23:02	+ BBS_PROJECT=sel4
    07-Aug-2020 14:23:02	+ BBS_REPO=seL4_tools
    07-Aug-2020 14:23:02	+ REPORT_TO_RUN=astyle
    07-Aug-2020 14:23:02	++ git rev-parse HEAD
    07-Aug-2020 14:23:02	+ COMMIT_ID=ccc38660229116bd26807c7012636c70e1c873f3
    07-Aug-2020 14:23:02	+ REPORT_KEY=astyle.report
    07-Aug-2020 14:23:02	+ set +x
    07-Aug-2020 14:23:02	+ report_file=astyle_report.json
    07-Aug-2020 14:23:02	+ annotations_file=astyle_annotations.json
    07-Aug-2020 14:23:02	+ rm -f astyle_report.json astyle_annotations.json
    07-Aug-2020 14:23:02	+ /scripts/bb_code_insights/astyle
    07-Aug-2020 14:23:04	+ echo 'Reporting the results to BitBucket, but doing so *secretly*, so people don'\''t see the access token'
    07-Aug-2020 14:23:04	Reporting the results to BitBucket, but doing so *secretly*, so people don't see the access token
    07-Aug-2020 14:23:04	+ set +x
    07-Aug-2020 14:23:04	+ '[' -f astyle_report.json ']'
    07-Aug-2020 14:23:04	+ cat astyle_report.json
    07-Aug-2020 14:23:04	+ '[' -f astyle_annotations.json ']'
    07-Aug-2020 14:23:04	{"data":[{"title":"Total number of files with incorrect style","value":0},{"title":"Total number of lines with incorrect style","value":0}],"createdDate":1596097395042,"key":"astyle.report","title":"Code style"}{"title": "Code style", "data": [{"title": "Total number of files with incorrect style", "value": 0}, {"title": "Total number of lines with incorrect style", "value": 0}]}
    07-Aug-2020 14:23:04	Finished task 'Run astyle code insights' with result: Success
    07-Aug-2020 14:23:04	Running post build plugin 'NCover Results Collector'
    07-Aug-2020 14:23:04	Running post build plugin 'Artifact Copier'
    07-Aug-2020 14:23:04	Running post build plugin 'npm Cache Cleanup'
    07-Aug-2020 14:23:04	Running post build plugin 'Clover Results Collector'
    07-Aug-2020 14:23:04	Running post build plugin 'Docker Container Cleanup'
    07-Aug-2020 14:23:04	Successfully removed working directory at '/var/atlassian/application-data/bamboo/xml-data/build-dir/151224373/CQ-SEL4TOOLSSTYLE154-SEL4TOOLSSTYLEPR'
    07-Aug-2020 14:23:04	Finalising the build...
    07-Aug-2020 14:23:04	Stopping timer.
    07-Aug-2020 14:23:04	Build CQ-SEL4TOOLSSTYLE154-SEL4TOOLSSTYLEPR-1 completed.
    07-Aug-2020 14:23:04	Running on server: post build plugin 'NCover Results Collector'
    07-Aug-2020 14:23:04	Running on server: post build plugin 'Build Hanging Detection Configuration'
    07-Aug-2020 14:23:04	Running on server: post build plugin 'Clover Delta Calculator'
    07-Aug-2020 14:23:04	Running on server: post build plugin 'Maven Dependencies Postprocessor'
    07-Aug-2020 14:23:04	All post build plugins have finished
    07-Aug-2020 14:23:04	Generating build results summary...
    07-Aug-2020 14:23:04	Saving build results to disk...
    07-Aug-2020 14:23:04	Store variable context...
    07-Aug-2020 14:23:04	Indexing build results...
    07-Aug-2020 14:23:04	Finished building CQ-SEL4TOOLSSTYLE154-SEL4TOOLSSTYLEPR-1.


This is the most I can report on right now, sorry!

**Pull request merged**